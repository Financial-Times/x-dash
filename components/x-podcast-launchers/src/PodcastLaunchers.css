.container {
	width: 100%;
}

.podcast-app-links__wrapper {
	list-style: none;
	padding: 0;
	margin: 0;
}

.podcast-app-link {
	width: 100%;
	margin-bottom: 8px;
}

.rss-url__wrapper {
	margin-top: 0px;
}

.rss-url__input {
	max-width: none;
}

.rss-url__copy-button {
	padding-left: 0px;
}

.rss-url__copy-span {
	-webkit-user-select: text;
	-khtml-user-select: text;
	-moz-user-select: text;
	-ms-user-select: text;
	-o-user-select: text;
	user-select: text;

	position: absolute;
	clip: rect(0 0 0 0);
	margin: -1px;
	border: 0;
	overflow: hidden;
	padding: 0;
	width: 1px;
	height: 1px;
	white-space: nowrap;
}
