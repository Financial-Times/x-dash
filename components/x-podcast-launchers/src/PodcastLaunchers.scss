:global {
	@import "~@financial-times/x-follow-button/dist/FollowButton";
}

.container {
	font-family: MetricWeb,sans-serif;
	width: 100%;
}

.heading {
	font-weight: 600;
	font-size: 18px;
	line-height: 20px;
	margin: 0 0 12px;
}

.podcast-app-links__wrapper {
	list-style: none;
	padding: 8px 0 0;
	margin: 0;
}

.podcast-app-link {
	width: 100%;
	margin-bottom: 8px;
}

.rss-url__wrapper {
	margin-top: 0px;
	margin-bottom: 24px;
}

.rss-url__input {
	max-width: none;
}

.rss-url__copy-button {
	padding-left: 0px;
}

.rss-url__copy-span {
	user-select: text;
	position: absolute;
	clip: rect(0 0 0 0);
	margin: -1px;
	border: 0;
	overflow: hidden;
	padding: 0;
	width: 1px;
	height: 1px;
	white-space: nowrap;
}
