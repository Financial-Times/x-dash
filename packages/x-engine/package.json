{
  "name": "@financial-times/x-engine",
  "version": "1.0.0-1",
  "description": "A consolidation library to render x- components with any compatible runtime.",
  "main": "src/server.js",
  "browser": "src/client.js",
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "assign-deep": "^0.4.7"
  },
  "engines": {
    "node": ">= 6.0.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
