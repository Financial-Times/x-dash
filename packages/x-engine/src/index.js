const path = require('path');
const deepGet = require('./concerns/deep-get');

const cwd = process.cwd();

// 1. try to load the application's package manifesr
const manifest = path.join(cwd, 'package.json');

let pkg;

try {
	pkg = require(manifest);
} catch(error) {
	throw new Error(`Could not load package.json from ${cwd}`);
}

// 2. if we have the manifest then find the engine configuration
const runtime = deepGet(pkg, 'x-dash.engine.server');

if (typeof runtime !== 'string') {
	throw new Error('x-engine requires a server runtime engine to be specified');
}

// 3. if this module is a linked dependency then resolve required runtime to CWD
const symlink = __dirname.startsWith(cwd) === false;
const dependency = symlink ? path.join(cwd, 'node_modules', runtime) : runtime;

let engine;

try {
	engine = require(dependency);
} catch (error) {
	// TODO: A friendlier error message
	throw new Error(`Failed to load ${runtime} from ${dependency}`);
}

// 4. if we've loaded the runtime then find it's create element factory function
const factory = typeof engine === 'function' ? engine : (engine.createElement || engine.h);

module.exports = factory;
