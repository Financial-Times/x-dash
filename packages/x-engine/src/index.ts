import { SFC, Attributes, ReactNode, ReactElement } from 'react';
import { join } from 'path';
import { readFileSync } from 'fs';
import deepGet from './concerns/deep-get';

// All components can extend a generic functional component without explicitly
// importing any React typings.
export type Component<P> = SFC<P>;

// Any engine we load should expose have a generic factory interface
export type Factory = <P>(
	type: Component<P> | string,
	props?: Attributes & P | null,
	...children: ReactNode[]
) => ReactElement<P> | string;

const manifest = join(process.cwd(), 'package.json');

const pkg = JSON.parse(readFileSync(manifest).toString());

const runtime = deepGet(pkg, 'x-dash.engine.server');

if (typeof runtime !== 'string') {
	throw new Error('x-engine requires a server runtime engine to be specified');
}

let engine;

try {
	engine = require('./engines/' + runtime) as Factory;
} catch (error) {
	// TODO: A friendlier error message
	throw new Error(`Failed to load "${runtime}"`);
}

export const h = engine;
