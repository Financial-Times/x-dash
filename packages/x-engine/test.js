const React = require('react');
const Teaser = require('./')

React.renderToString(Teaser({
	type: 'article',
	id: '',
	url: '#',
	title: 'Inside charity fundraiser where hostesses are put on show',
	titleVariant: '',
	standfirst: 'FT investigation finds groping and sexual harassment at secretive black-tie dinner',
	standfirstVariant: '',
	publishedDate: '2018-01-23T15:07:00.000Z',
	firstPublishedDate: '2018-01-23T13:53:00.000Z',
	conceptPrefix: '',
	conceptSuffix: '',
	concept: {
		url: '#',
		prefLabel: 'FT Investigations'
	},
	alternativeConcept: {
		url: '#',
		prefLabel: ''
	},
	image: {
		url: 'http://prod-upp-image-read.ft.com/a25832ea-0053-11e8-9650-9c0ad2d7c5b5',
		width: 2048,
		height: 1152,
		aspectRatio: 0.5625
	},
	premium: false
}))
