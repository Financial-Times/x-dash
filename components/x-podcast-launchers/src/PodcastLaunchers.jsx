import { h, Component } from '@financial-times/x-engine';
import generateAppLinks from './generate-app-links';
import generateRSSUrl from './generate-rss-url';
import styles from './PodcastLaunchers.css';
import copyToClipboard from './copy-to-clipboard';

const basicButtonStyles = [
	'o-buttons',
	'o-buttons--primary',
	'o-buttons--big'
].join(' ');

const podcastAppLinkStyles = [
	basicButtonStyles,
	styles['podcast-app-link']
].join(' ');

const rssUrlWrapperStyles = [
	'o-forms__affix-wrapper',
	styles['rss-url__wrapper']
].join(' ');

const rssUrlInputStyles = [
	'o-forms__text',
	styles['rss-url__input']
].join(' ');

const rssUrlCopyButtonWrapperStyles = [
	'o-forms__suffix',
	styles['rss-url__copy-button']
].join(' ');

function mapConceptToAcastSeries (/** concept */) {
	return 'ft-test';
}

class PodcastLaunchers extends Component {
	constructor(props) {
		super(props);
		this.state = {
			rssUrl: ''
		}
	}

	componentDidMount() {
		const { seriesConcept, acastRSSHost, acastAccessToken } = this.props;
		const acastSeries = mapConceptToAcastSeries(seriesConcept);
		if (acastSeries) {
			this.setState({
				rssUrl: generateRSSUrl(acastSeries, acastRSSHost, acastAccessToken)
			});
		}

	}

	render() {
		const { rssUrl } = this.state;
		return (
			<div className={styles['container']}>
				{rssUrl && (
					<ul className={styles['podcast-app-links__wrapper']}>
					{generateAppLinks(rssUrl).map(({ name, url }) => (
						<li key={name}>
							<a href={url} className={podcastAppLinkStyles}>{name}</a>
						</li>
					))}
					</ul>
				)}
				<div className={rssUrlWrapperStyles}>
					<input className={rssUrlInputStyles} value={rssUrl} type='text' readOnly/>
					<div className={rssUrlCopyButtonWrapperStyles}>
						<button className={basicButtonStyles} onClick={copyToClipboard}>Copy RSS</button>
					</div>
				</div>
			</div>
		)
	}
}

export { PodcastLaunchers };
