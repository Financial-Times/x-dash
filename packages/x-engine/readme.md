# x-engine

This module is a consolidation library to render `x-dash` components with any compatible runtime.

## Installation

This module is supported on Node 12 and is distributed on npm.

```bash
npm install -S @financial-times/x-engine
```

You'll also need to install your chosen runtime and any related dependencies. Some compatible runtimes are:

- [Hyperapp](https://github.com/hyperapp/hyperapp)<sup>\*</sup>
- [Hyperons](https://github.com/i-like-robots/hyperons)
- [Inferno](https://infernojs.org/)
- [Nerv](https://github.com/NervJS/nerv)
- [Preact](https://preactjs.com/)
- [Rax](https://alibaba.github.io/rax/)
- [React](https://reactjs.org/)
- [VDO](https://github.com/DylanPiercey/vdo)

\* Usage of Hyperapp depends on a small modification to higher-order components to accept `children` as a second argument rather than receiving them appended to `props`.

## Configuration

To start you must specify your runtime configuration within `package.json`. This instructs `x-engine` which modules to load for your environment. You may specify different runtimes for server and browser rendering depending on your needs.

This module includes several presets for popular tools and frameworks including React, Preact, and Hyperons. For example to use Hyperons on the server and Preact in the browser you may use the name of their preset:

```json
{
	"x-dash": {
		"engine": {
			"server": "hyperons",
			"browser": "preact"
		}
	}
}
```

But if your chosen tool does not have a preset then the configuration for it can be provided with the expanded configuration format. The example below shows how to load the VDO module and use its `createElement` factory function<sup>\*</sup>:

```json
{
	"x-dash": {
		"engine": {
			"server": {
				"runtime": "vdo",
				"factory": "createElement"
			}
		}
	}
}
```

\* A JSX factory function is a variadic function<sup>†</sup> with the signature `fn(element, properties, ...children)`, examples include `React.createElement` and `Preact.h`. See the [FAQ section](#faq) for more information.

† Variadic means that the function accepts a variable number of arguments. The `...` before the last arguments name is a rest parameter, meaning it will collect "the rest" of the arguments.

## Rendering

### Server-side

If your chosen runtime factory returns a string then you can pass properties to the component and immediately use the returned value:

```js
const { Teaser } = require('@financial-times/x-teaser');

app.get('/teaser', (request, response) => {
	const properties = { … };
	response.send(Teaser(properties));
});
```

But if your factory method returns a framework-specific intermediary code then you'll need to load the specified render method to render the node into a string or stream:

```js
const { render } = require('@financial-times/x-engine');
const { Teaser } = require('@financial-times/x-teaser');

app.get('/teaser', (request, response) => {
	const properties = { … };
	const nodes = Teaser(properties);
	response.send(render(nodes));
});
```

### Client-side

To use components on the client-side you will first need to add the `x-engine` plugin to your Webpack configuration file. Under the hood this uses [`DefinePlugin`](https://webpack.js.org/plugins/define-plugin/) to wire up your chosen runtime.

```js
// webpack.config.js
const xEngine = require('@financial-times/x-engine/src/webpack');

module.exports = {
	plugins: [
		xEngine()
	]
};
```

You can then install and use `x-` components in your client-side code:

```jsx
import { h } from '@financial-times/x-engine';
import { Teaser } from '@financial-times/x-teaser';

export default const TeaserList = (props) => (
	<ul class="TeaserList">
		{props.items.map((item) => (
			<li className="TeaserList-Item">
				<Teaser {...item} layout="small" showImage={true} />
			</li>
		))}
	</ul>
);
```

#### Client-side through n-ui
In order to configure n-ui to compile x-dash based `.jsx` files you will also need to add the following to `n-ui-build.config.js`.

```js
// n-ui-build.config.js
const xEngine = require('@financial-times/x-engine/src/webpack');

module.exports = {
	plugins: [
		xEngine()
	],
	pragma: 'h'
};
```

You will also need to import the factory function from `x-engine` in each `.jsx` file.

```javascript
import { h } from '@financial-times/x-engine'
```

## FAQ

### This sounds complicated… is it a magic black box?

There is no magic. The source code for the server-side integration is less than 60 lines of unexciting code. The Webpack plugin for client-side usage is even smaller.

### What is a "factory function"?

A factory function is a variadic function with the signature `fn(element, properties, ...children)`, examples include `React.createElement` and `Preact.h`. It may return the framework's representation of a HTML node or a formatted string depending on the runtime you're using.

### Which runtime should I use?

Whichever one you want! React, Preact, Rax, and Nerv are all largely compatible with one another. If you don't want the overhead of a framework, or are rendering static HTML, then it's worth investigating the Hyperons or VDO modules.

### What about Hyperscript?

Hyperscript currently only supports passing a tag name (a string) as the first argument. This limitation means you cannot currently reference components inside other components.
