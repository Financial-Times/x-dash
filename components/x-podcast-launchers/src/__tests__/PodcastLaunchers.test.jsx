import { h } from '@financial-times/x-engine';
import { brand } from '@financial-times/n-concept-ids';
import renderer from 'react-test-renderer';
jest.mock('../PodcastLaunchers.scss', () => ({}));
import { PodcastLaunchers } from '../PodcastLaunchers';


const acastRSSHost = 'https://acast.access';
const acastAccessToken = '123-abc';

describe('PodcastLaunchers', () => {
	it('should hide itself if an RSS URL could not be generated', () => {
		expect(
			renderer.create(<PodcastLaunchers conceptId='123-abc' {...{acastRSSHost, acastAccessToken}} />).toJSON()
		).toMatchSnapshot();
	});

	it('should render the app links based on concept Id', () => {
		expect(
			renderer.create(<PodcastLaunchers conceptId={brand.rachmanReviewPodcast} {...{acastRSSHost, acastAccessToken}} />).toJSON()
		).toMatchSnapshot();
	});
})
