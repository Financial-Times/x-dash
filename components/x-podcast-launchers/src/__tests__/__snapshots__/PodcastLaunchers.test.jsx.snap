// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PodcastLaunchers should hide itself if an RSS URL could not be generated 1`] = `""`;

exports[`PodcastLaunchers should render the app links based on concept Id 1`] = `
<div
  data-trackable="podcast-launchers"
>
  <h2>
    Subscribe via your installed podcast app
  </h2>
  <ul>
    <li
      className={null}
    >
      <a
        data-trackable="apple-podcasts"
        href="podcast://acast.access/rss/therachmanreview/123-abc"
      >
        Apple Podcasts
      </a>
    </li>
    <li
      className={null}
    >
      <a
        data-trackable="overcast"
        href="overcast://x-callback-url/add?url=https://acast.access/rss/therachmanreview/123-abc"
      >
        Overcast
      </a>
    </li>
    <li
      className={null}
    >
      <a
        data-trackable="pocket-casts"
        href="pktc://subscribe/acast.access/rss/therachmanreview/123-abc"
      >
        Pocket Casts
      </a>
    </li>
    <li
      className={null}
    >
      <a
        data-trackable="podcast-addict"
        href="podcastaddict://acast.access/rss/therachmanreview/123-abc"
      >
        Podcast Addict
      </a>
    </li>
    <li
      className={null}
    >
      <a
        data-trackable="acast"
        href="acast://subscribe/https://acast.access/rss/therachmanreview/123-abc"
      >
        Acast
      </a>
    </li>
    <li>
      <span
        className="  "
      >
        <input
          readOnly={true}
          type="text"
          value="https://acast.access/rss/therachmanreview/123-abc"
        />
        <button
          data-trackable="copy-rss"
          data-url="https://acast.access/rss/therachmanreview/123-abc"
          onClick={[Function]}
          type="button"
        >
          Copy RSS
        </button>
      </span>
    </li>
  </ul>
  <div
    className="podcast-launchers__no-app-wrapper "
  >
    <h2>
      Canâ€™t see your podcast app?
    </h2>
    <p>
      Get updates for new episodes
    </p>
    <form
      action="/__myft/api/core/followed/concept/e0f2acb4-4177-436d-a783-b8c80ec2a6ac?method=put"
      data-concept-id="e0f2acb4-4177-436d-a783-b8c80ec2a6ac"
      method="GET"
      onSubmit={[Function]}
    >
      <button
        aria-label="Add undefined to myFT"
        aria-pressed="false"
        className="main_button__3Mk67"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Add to myFT",
          }
        }
        data-concept-id="e0f2acb4-4177-436d-a783-b8c80ec2a6ac"
        data-trackable="follow"
        data-trackable-context-messaging={null}
        title="Add undefined to myFT"
        type="submit"
      />
    </form>
  </div>
</div>
`;
